### PR TITLE
fix: handle multi-entry dictionary after grammar statement

### LINKED ISSUE #73: Multi-entry dict not processed correctly
When followed by dictionary with more than one entry in a list, a grammar statement is removed by mistake.

The following configuration illustrate the issue:

```
foo:
- on amd64 to riscv64:
   - bar1: baz1
- on amd64 to arm64:
   - bar2: baz2
- bar3: baz3
  bar4: baz4
```

When building for `riscv64`, the first statement (the `on amd64 to riscv64` section) is finalized when dealing with the second one. Then the second one (the `on amd64 to arm64` section) is excluded (not matching the target arch). So this is producing the expected primitives.

When building for `arm64`, the first statement is finalized (and later ignored) when dealing with the second one. But the second one is removed by mistake in `_parse_section_dictionary()` when looping over entries in the last dictionary (specifically when handling `bar4: baz4`). So the only resulting primitive will be the last element of the list.

See  https://github.com/canonical/craft-grammar/commit/67816503fd77f033d27276720c14d621cd537009 showing a broken test.

### PR DESCRIPTION
**Pull request by upils**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----

Fixes #73

### CODE REVIEW

File: craft_grammar/_processor.py [hunk 1]
```diff
@@ -253,12 +253,21 @@ def _parse_section_dictionary(
...
                 # the current statement is complete and this section
                 # is a primitive to be collected.
-                finalized_statement = statement
+                finalized_statement = _finalize(statement, finalized_statement)
                 statement = None
 
         return statement, finalized_statement
 
 
+def _finalize(
```
Comment (tigarmo) [inline craft_grammar/_processor.py:262]: The change looks conceptually fine to me, but I worry about unintended breakages. Can you run Snapcraft's tests with this branch?

**Review by lengau — APPROVED** [review]
Weird edge case, but LGTM!

File: craft_grammar/_processor.py (hunk 1 above)
Comment (upils) [reply craft_grammar/_processor.py:262]: I ran Snapcraft's tests with the branch and found no failure related to grammar.

**Review by tigarmo — APPROVED** [review]
